pkg/operator/csr/csr_approver.go

```diff
@@ -0,0 +1,275 @@
+		case CSRApproved:
+			return c.approveCSR(ctx, csrCopy, syncCtx.Recorder())
+		case CSRNoOpinion:
+			fallthrough
+		default:
+			return nil
+		}
+	}
+
+	// no matching CSRs were found
+	return nil
+}
+
+func (c *csrApproverController) denyCSR(ctx context.Context, csrCopy *certapiv1.CertificateSigningRequest, reason, message string, eventsRecorder events.Recorder) error {
+	csrCopy.Status.Conditions = append(csrCopy.Status.Conditions,
+		certapiv1.CertificateSigningRequestCondition{
+			Type:    certapiv1.CertificateDenied,
+			Status:  corev1.ConditionTrue,
+			Reason:  reason,
+			Message: message,
+		},
+	)
+
+	eventsRecorder.Eventf("CSRDenial", "The CSR %q has been denied", csrCopy.Name)
+	_, err := c.csrClient.UpdateApproval(ctx, csrCopy.Name, csrCopy, v1.UpdateOptions{})
+	return err
+}
+
+func (c *csrApproverController) approveCSR(ctx context.Context, csrCopy *certapiv1.CertificateSigningRequest, eventsRecorder events.Recorder) error {
+	csrCopy.Status.Conditions = append(csrCopy.Status.Conditions,
+		certapiv1.CertificateSigningRequestCondition{
+			Type:    certapiv1.CertificateApproved,
+			Status:  corev1.ConditionTrue,
+			Reason:  "AutoApproved",
+			Message: fmt.Sprintf("Auto-approved CSR %q", csrCopy.Name),
+		})
+
+	eventsRecorder.Eventf("CSRApproval", "The CSR %q has been approved", csrCopy.Name)
+	_, err := c.csrClient.UpdateApproval(ctx, csrCopy.Name, csrCopy, v1.UpdateOptions{})
+	return err
+}
+
+func getCertApprovalCondition(status *certapiv1.CertificateSigningRequestStatus) (approved bool, denied bool) {
+	for _, c := range status.Conditions {
+		if c.Type == certapiv1.CertificateApproved {
+			approved = true
+		}
+		if c.Type == certapiv1.CertificateDenied {
+			denied = true
+		}
+	}
+	return
+}
+
+type ServiceAccountApprover struct {
+	saGroups        sets.String // saGroups is the set of groups for the SA expected to have created the CSR
+	saName          string
+	expectedSubject string
+}
+
+// ServiceAccountApprover approves CSRs with a given subject issued by the provided service account
+func NewServiceAccountApprover(saNamespace, saName, expectedSubject string) *ServiceAccountApprover {
+	return &ServiceAccountApprover{
+		saName:          serviceaccount.MakeUsername(saNamespace, saName),
+		saGroups:        sets.NewString(append(serviceaccount.MakeGroupNames(saNamespace), "system:authenticated")...),
+		expectedSubject: expectedSubject,
+	}
+}
+
+func (a *ServiceAccountApprover) Approve(csrObj *certapiv1.CertificateSigningRequest, x509CSR *x509.CertificateRequest) (approvalStatus CSRApprovalDecision, denyReason string, err error) {
+	if csrObj == nil || x509CSR == nil {
+		return CSRDenied, "Error", fmt.Errorf("received a 'nil' CSR")
+	}
+
+	if csrObj.Spec.Username != a.saName {
+		return CSRDenied, fmt.Sprintf("CSR %q was created by an unexpected user: %q", csrObj.Name, csrObj.Spec.Username), nil
+	}
+
+	if csrGroups := sets.NewString(csrObj.Spec.Groups...); !csrGroups.IsSuperset(a.saGroups) {
```

Comment:
without groups, you need to deny anything specifying a group.

Reply:
I'm going to allow specifying additional groups and make this check and `Equal()`, then